fdbserver/workloads/ReportConflictingKeys.actor.cpp

```diff
@@ -0,0 +1,166 @@
+		m.push_back(PerfMetric("Conflicts/sec", conflicts.getValue() / testDuration, true));
+        m.push_back(retries.getMetric());
+    }
+
+    // disable the default timeout setting
+	double getCheckTimeout() override {return std::numeric_limits<double>::max();}
+
+    // Copied from tester.actor.cpp, added parameter to determine the key's length
+    Key keyForIndex( int n) {
+        double p = (double) n / nodeCount;
+        int paddingLen = keyBytes - 16 - keyPrefix.size();
+        // left padding by zero
+        return StringRef(format("%0*llx", paddingLen, *(uint64_t*)&p)).withPrefix(keyPrefix);
+    }
+
+    void addRandomReadConflictRange(ReadYourWritesTransaction* tr, std::vector<KeyRange>& readConflictRanges) {
+        int startIdx, endIdx;
+        Key startKey, endKey;
+        while (deterministicRandom()->random01() < addReadConflictRangeProb) {
+            startIdx = deterministicRandom()->randomInt(0, nodeCount);
+            endIdx = deterministicRandom()->randomInt(startIdx, nodeCount);
+            startKey = keyForIndex(startIdx);
+            endKey = keyForIndex(endIdx);
+            tr->addReadConflictRange(KeyRangeRef(startKey, endKey));
+            readConflictRanges.push_back(KeyRangeRef(startKey, endKey));
+        }
+    }
+
+    void addRandomWriteConflictRange(ReadYourWritesTransaction* tr) {
+        int startIdx, endIdx = 0;
+        Key startKey, endKey;
+        while (deterministicRandom()->random01() < addWriteConflictRangeProb) {
+            startIdx = deterministicRandom()->randomInt(endIdx, nodeCount);
+            endIdx = deterministicRandom()->randomInt(startIdx, nodeCount);
+            startKey = keyForIndex(startIdx);
+            endKey = keyForIndex(endIdx);
+            tr->addWriteConflictRange(KeyRangeRef(startKey, endKey));
+        }
+    }
+
+    ACTOR Future<Void> conflictingClient(Database cx, ReportConflictingKeysWorkload* self, double delay, int actorIndex) {
+
+        state ReadYourWritesTransaction tr(cx);
+        state double lastTime = now();
+        state std::vector<KeyRange> readConflictRanges;
+
+        loop {
+            try {
+                // used for throttling
+                wait(poisson(&lastTime, delay));
+                if (self->reportConflictingKeys)
+                    tr.setOption(FDBTransactionOptions::REPORT_CONFLICTING_KEYS);
+                self->addRandomReadConflictRange(&tr, readConflictRanges);
+                self->addRandomWriteConflictRange(&tr);
+                wait(tr.commit());
+                ++self->xacts;
+            } catch (Error& e) {
+                state bool isConflict = false;
+                if (e.code() == error_code_operation_cancelled)
+					throw;
+                else if (e.code() == error_code_not_committed) {
+                    ++self->conflicts;
+                    isConflict = true;
+                }
+                wait(tr.onError(e));
+                // check API correctness
```

Comment:
This looks good - I still want to get buy-in for what the exact semantics/contract of this feature should be from the layers folks before shipping though

Reply:
I'm now thinking that we want to report a key range which intersects both the read conflict range of this transaction and the write conflict range of some other transaction, if recovering the exact write conflict range of the other transaction isn't feasible. Does that sound reasonable?

Reply:
This makes sense. Has changed to narrow the returned `KeyRange` to be the intersection of original `readConflictRange` and some `writeConflictRange`.